docker/Dockerfile.armv7-unknown-linux-gnueabihf

```diff
@@ -1,4 +1,4 @@
-FROM ubuntu:16.04
+FROM ubuntu:20.04
```

Comment:
This image, and probably more, reference a `cc` that is not in path: `arm-linux-gnueabihf-gcc` does not exist.

Reply:
discussed in matrix

https://matrix.to/#/!ZFqvHUjUqgONNObpSJ:matrix.org/$fNBbYZ_-RWoY5pNjbyMKkLKFedrfqYahsuMZDZr_dhI?via=matrix.org&via=matrix.questionable.software&via=rduce.org

Reply:
this is wrong, it does exist

```
$ cargo build-docker-image armv7-unknown-linux-gnueabihf --tag ubuntu20
[..snip]
 => => writing image sha256:fa5d48f3e9c18589f373cdeece4b611c1d5f3dcea31cdc2d25ce0c2d575f4b23                                                                                                                    0.0s
 => => naming to ghcr.io/cross-rs/armv7-unknown-linux-gnueabihf:ubuntu20   
$ docker run --rm -it ghcr.io/cross-rs/armv7-unknown-linux-gnueabihf:ubuntu20
root@df34d3813315:/# arm-linux-gnueabihf-
arm-linux-gnueabihf-addr2line     arm-linux-gnueabihf-dwp           arm-linux-gnueabihf-gcc-ar        arm-linux-gnueabihf-gcov          arm-linux-gnueabihf-gprof         arm-linux-gnueabihf-objdump
arm-linux-gnueabihf-ar            arm-linux-gnueabihf-elfedit       arm-linux-gnueabihf-gcc-ar-9      arm-linux-gnueabihf-gcov-9        arm-linux-gnueabihf-ld            arm-linux-gnueabihf-ranlib
arm-linux-gnueabihf-as            arm-linux-gnueabihf-g++           arm-linux-gnueabihf-gcc-nm        arm-linux-gnueabihf-gcov-dump     arm-linux-gnueabihf-ld.bfd        arm-linux-gnueabihf-readelf
arm-linux-gnueabihf-c++filt       arm-linux-gnueabihf-g++-9         arm-linux-gnueabihf-gcc-nm-9      arm-linux-gnueabihf-gcov-dump-9   arm-linux-gnueabihf-ld.gold       arm-linux-gnueabihf-size
arm-linux-gnueabihf-cpp           arm-linux-gnueabihf-gcc           arm-linux-gnueabihf-gcc-ranlib    arm-linux-gnueabihf-gcov-tool     arm-linux-gnueabihf-nm            arm-linux-gnueabihf-strings
arm-linux-gnueabihf-cpp-9         arm-linux-gnueabihf-gcc-9         arm-linux-gnueabihf-gcc-ranlib-9  arm-linux-gnueabihf-gcov-tool-9   arm-linux-gnueabihf-objcopy       arm-linux-gnueabihf-strip
````

but, installing binutils removes it

```
root@df34d3813315:/# apt-get install --no-install-recommends binutils-arm-linux-gnueabihf:armhf binutils:armhf libclang-10-dev:armhf clang-10:armhf
Reading package lists... Done
Building dependency tree       
Reading state information... Done
The following packages were automatically installed and are no longer required:
  binutils-common binutils-x86-64-linux-gnu cpp cpp-9 cpp-9-arm-linux-gnueabihf cpp-arm-linux-gnueabihf gcc-10-cross-base gcc-9-arm-linux-gnueabihf-base gcc-9-base gcc-9-cross-base lib32gcc-s1 lib32stdc++6
  libasan5 libasan5-armhf-cross libatomic1 libatomic1-armhf-cross libbinutils libbsd0 libc6-dev libc6-i386 libcc1-0 libcrypt-dev libctf-nobfd0 libctf0 libedit2 libgc1c2 libgcc-9-dev libgcc-9-dev-armhf-cross
  libgcc-s1-armhf-cross libgomp1 libgomp1-armhf-cross libisl22 libitm1 libllvm10 liblsan0 libmpc3 libmpfr6 libobjc-9-dev libobjc4 libquadmath0 libstdc++-9-dev libstdc++-9-dev-armhf-cross libstdc++6-armhf-cross
  libtsan0 libubsan1 libubsan1-armhf-cross linux-libc-dev
Use 'apt autoremove' to remove them.
The following additional packages will be installed:
  binutils-common:armhf gcc-9-base:armhf libasan5:armhf libatomic1:armhf libbinutils:armhf libbsd0:armhf libclang-common-10-dev:armhf libclang-cpp10:armhf libclang1-10:armhf libctf-nobfd0:armhf libctf0:armhf
  libedit2:armhf libffi7:armhf libgc1c2:armhf libgcc-9-dev:armhf libgomp1:armhf libllvm10:armhf libobjc-9-dev:armhf libobjc4:armhf libstdc++-9-dev:armhf libstdc++6:armhf libubsan1:armhf
Suggested packages:
  binutils-doc:armhf clang-10-doc:armhf libstdc++-9-doc:armhf
Recommended packages:
  llvm-10-dev:armhf python3:armhf libomp-10-dev:armhf
The following packages will be REMOVED:
  binutils binutils-arm-linux-gnueabihf g++ g++-9 g++-9-arm-linux-gnueabihf g++-arm-linux-gnueabihf gcc gcc-9 gcc-9-arm-linux-gnueabihf gcc-arm-linux-gnueabihf libclang-10-dev libclang-common-10-dev libclang-dev
  libclang1-10 libtool
The following NEW packages will be installed:
  binutils:armhf binutils-arm-linux-gnueabihf:armhf binutils-common:armhf clang-10:armhf gcc-9-base:armhf libasan5:armhf libatomic1:armhf libbinutils:armhf libbsd0:armhf libclang-10-dev:armhf
  libclang-common-10-dev:armhf libclang-cpp10:armhf libclang1-10:armhf libctf-nobfd0:armhf libctf0:armhf libedit2:armhf libffi7:armhf libgc1c2:armhf libgcc-9-dev:armhf libgomp1:armhf libllvm10:armhf
  libobjc-9-dev:armhf libobjc4:armhf libstdc++-9-dev:armhf libstdc++6:armhf libubsan1:armhf
0 upgraded, 26 newly installed, 15 to remove and 12 not upgraded.
Need to get 58.0 MB of archives.
After this operation, 35.7 MB disk space will be freed.
```

this is just a pitfall, installing `clang:armhf` etc is wrong, iirc, it's possible to set some packages as forbidden, could we add all the packages that would uninstall the build tools to such a banlist?

Reply:
Ok we can use `/etc/apt/preferences.d` with pinning to forbid the installation of these packages:

**cat /etc/apt/preferences.d/clang**

```text
Package: clang-10:armhf
Pin: release *
Pin-Priority: -1
```

And this gives:

```bash
$ apt-get install clang-10:armhf
Reading package lists... Done
Building dependency tree
Reading state information... Done
Package clang-10:armhf is not available, but is referred to by another package.
This may mean that the package is missing, has been obsoleted, or
is only available from another source

E: Package 'clang-10:armhf' has no installation candidate
```